src/socket_base.cpp

```diff
@@ -1414,6 +1414,7 @@ void zmq::socket_base_t::update_pipe_options(int option_)
         for (pipes_t::size_type i = 0; i != pipes.size(); ++i)
         {
             pipes[i]->set_hwms(options.rcvhwm, options.sndhwm);
+            pipes[i]->set_peer_hwms(options.sndhwm, options.rcvhwm);
```

Comment:
@bluca You are right, I had not considered inproc case before.
I think there are two solution:
[1] disable update hwm after connect, since this is not a common usecase
[2] add a command type in command.hpp to support modify pipe by it's owner thread. If we detect the protocol is inproc, we can set the command, otherwises, set_peer_hwm directly.